[DRAFT] feat: granule links via event subscription

ceholden opened the pull request:
> [!IMPORTANT]
> This is a draft! I'm hoping to ensure unit and integration tests are working and finish self-reviewing before tagging anyone.
> As such the PR description is incomplete 😜 

## What I am changing

See https://github.com/NASA-IMPACT/hls_development/issues/300

## How I did it

## How you can test it

ceholden commented on lambdas/link_fetcher/manage_subscription.py:
This is modified from the example ESA linked so I've included attribution in the header. The code example covered getting an auth token and basic create/list/delete subscriptions operations. Happy to change anything about this

I figured we could use this to create the subscription manually for now, but it's not strictly needed (not used in CI or deployment). The subscriptions API can also update existing subscriptions (i.e., to rotate credentials, to pause/resume, or to point to another API endpoint) but I haven't added that yet without a use in mind

ciaransweet commented on lambdas/link_fetcher/README.md:
```diff
@@ -1,14 +1,23 @@
+* Subscriptions API ("event based")
+
+We wish to migrate from a polling to an event driven method of link
+fetchiung. During the transition period, this subdirectory handles both
```
```suggestion
fetching. During the transition period, this subdirectory handles both
```

chuckwondo reviewed (approved):
@ceholden, fantastic work on this PR!

I didn't find anything to block approval. I just found minor things, and had some questions for clarification, and some "food for thought."

ceholden commented:
Thanks @chuckwondo! I'll get to your feedback later today or first thing tomorrow but it looks like great input, really appreciate your close review!

ceholden commented on integration_tests/test_link_push_subscription.py:
```diff
@@ -0,0 +1,150 @@
+        / "data"
+        / "push-granule-created-s2-n1.json"
+    )
+    with data.open() as src:
+        payload = json.load(src)
```
Oooh very nice! I didn't catch this addition in Py3.5 but now I'll be using `[read|write]_[text|bytes]`, thanks for the suggestion

ceholden commented on lambdas/link_fetcher/app/subscription_endpoint.py:
```diff
@@ -0,0 +1,209 @@
+                "Could not retrieve ESA subscription credentials from Secrets Manager"
+            ) from ex
+
+        return cls(
+            notification_username=secret["notification_username"],
+            notification_password=secret["notification_password"],
+        )
```
Definitely! Great callout, I'll fix this in a commit

ceholden commented on lambdas/link_fetcher/tests/test_subscription_endpoint.py:
```diff
@@ -0,0 +1,288 @@
+    We freezse time
+    """
+    # freeze to same day as publication date
+    with freeze_time(event_s2_created["value"]["PublicationDate"]):
```
I think I saw this pattern in one of your PRs, and +1 this seems like a nice approach for the reasons you laid out.

For datetimes I've found bugs in way too many places from people hard coding definitions of `now`, for instance when containers live long enough to celebrate the new year 🎉 so the defined "current year" becomes incorrect 😭. The suggestion you have of using a `Callable[[], datetime]` avoids this and reminds me of "dependency inject vs mock/patch", and where feasible I tend to be a fan of the former... so to summarize this sounds great!

ceholden commented on lambdas/link_fetcher/tests/test_subscription_endpoint.py:
```diff
@@ -0,0 +1,288 @@
+            yield config
+
+    def test_local_from_ssm(self, monkeypatch, endpoint_config_secret: EndpointConfig):
+        monkeypatch.setenv("STAGE", endpoint_config_secret.stage)
+        config = EndpointConfig.load_from_secrets_manager(endpoint_config_secret.stage)
```
+1, and this "why am I monkeypatching `STAGE`?" should've made me realize I was missing the pass through from this classmethod! Thanks for the suggestion

ceholden commented on lambdas/link_fetcher/app/subscription_endpoint.py:
```diff
@@ -0,0 +1,209 @@
+        ssm_client: "SSMClient",
+    ) -> str:
+        """Return the endpoint URL stored in SSM parameter store"""
+        param_name = "/integration_tests/{stage}/link_subscription_endpoint_url".format(
+            stage=self.stage
+        )
```
Good point! I had been using this as part of the `manage_subscriptions.py` script but that's the only place. I can move this lookup into that script

ceholden commented on cdk/downloader_stack.py:
```diff
@@ -286,6 +286,59 @@ def __init__(
+            ],
+            memory_size=200,
+            timeout=Duration.minutes(15),
+            runtime=aws_lambda.Runtime.PYTHON_3_11,
```
+1 it looks specific to the Python-specific `aws_lambda_python_alpha` version of CDK. I was wondering how you got the GIBS stuff working on Py3.12!

Specifically this Python-specific Lambda library installs Pipenv and Poetry as part of setup, and there's some `pipenv` <> `py3.12` interaction that is going badly, see https://github.com/aws/aws-cdk/issues/30170#issuecomment-2137950708

They're tracking the issue but there isn't a resolution yet. We could avoid using `aws_lambda_python_alpha` if we want to export the `Pipfile.lock` to a `requirements.txt` ahead of the CDK deploy stage, but I'm hoping there'll be a fix upstream instead of taking on that work

ceholden commented:
Hey @chuckwondo, thanks for your reviews! I think I addressed your feedback and this PR is now up to date with `main` after fixing some merge conflicts with the work related to updating for ESA's change in search query offset/limit. Would you mind re-reviewing?

I won't turn on the pushes to this endpoint until after some more acceptance testing in another stage deployment that writes to a non-production bucket, but it'd be nice to get it in the main Git branch in case there are other changes that come up.

chuckwondo commented on lambdas/link_fetcher/tests/test_subscription_endpoint.py:
```diff
@@ -0,0 +1,289 @@
+import json
```
trivial type hinting adjustment (see related comment against a fixture)

```suggestion
import json
from collections.abc import Iterator
```

chuckwondo commented on lambdas/link_fetcher/tests/test_subscription_endpoint.py:
```diff
@@ -0,0 +1,289 @@
+        assert config.notification_password == "baz"
+
+    @pytest.fixture
+    def endpoint_config_secret(self) -> EndpointConfig:
```
```suggestion
    def endpoint_config_secret(self) -> Iterator[EndpointConfig]:
```

ceholden replied:
Good catch!

chuckwondo commented on lambdas/link_fetcher/tests/test_subscription_endpoint.py:
```diff
@@ -0,0 +1,289 @@
+        This is also distinguished by using a "recent" event notification that works
+        by freezing/rewinding time to acquisition date for the test duration.
+        """
+        # we can't inject an older "utc_now" into `process_notification` so we have to
+        # patch `datetime.now()` with freezegun
+        with freeze_time(event_s2_created["value"]["PublicationDate"]):
+            resp = test_client.post(
+                "/events",
+                json=event_s2_created,
+                auth=(
+                    self.endpoint_config.notification_username,
+                    self.endpoint_config.notification_password,
+                ),
+            )
```
Actually, I believe we _can_ inject an older "utc_now" by doing the following, along with the suggestion I made for `subscription_endpoint.py`.

However, feel free to skip this (along with the corresponding suggestion on `subscription_endpoint.py`), although I'm curious to see if this would work.

```suggestion
        setattr(
            test_client.app,
            "now_utc",
            lambda: datetime.fromisoformat(
                event_s2_created["value"]["PublicationDate"]
            ),
        )
        resp = test_client.post(
            "/events",
            json=event_s2_created,
            auth=(
                self.endpoint_config.notification_username,
                self.endpoint_config.notification_password,
            ),
        )
```

chuckwondo commented on lambdas/link_fetcher/app/subscription_endpoint.py:
```diff
@@ -0,0 +1,211 @@
+            raise HTTPException(status_code=401, detail="Unauthorized")
+
+        # process notification
+        process_notification(
+            notification=notification,
+            accepted_tile_ids=accepted_tile_ids,
+            session_maker=session_maker,
+        )
```
```suggestion
        process_notification(
            notification=notification,
            accepted_tile_ids=accepted_tile_ids,
            session_maker=session_maker,
            now_utc=getattr(
                request.app, "now_utc", lambda: datetime.now(tz=timezone.utc)
            ),
         )
```

ceholden replied:
Nice idea! Appreciate you pushing for this sort of thing. I ended up doing this a little differently, but still was able to inject the "what time is it?" dependency

Instead of storing the lookup function on the FastAPI application I added `now_utc` as a parameter to the `build_app` function. To me this felt like a little bit more direct than setting/getting an attribute off of the `.app` itself, and I think it mirrors how the `config` is injected into the FastAPI app builder

I could not remove `freezegun` as a dependency because a handful of tests related to the scheduled search-based link fetcher code still use it. This was the commit with the changes, https://github.com/NASA-IMPACT/hls-sentinel2-downloader-serverless/pull/44/commits/23ed64f9f27eccb512725ebe859ab66bc2b1cb99

chuckwondo reviewed (approved):
Great work, as always. Just a few "food for thought" comments you can choose to ignore. However, if you do use my comments regarding freeze_time, and they work, you should then be able to drop the freezegun dependency.

ceholden commented on lambdas/link_fetcher/app/subscription_endpoint.py:
```diff
@@ -0,0 +1,211 @@
+            f"/hls-s2-downloader-serverless/{self.stage}/link_subscription_endpoint_url"
+        )
+        result = ssm_client.get_parameter(Name=param_name)
+        if (value := result["Parameter"].get("Value")) is None:
+            raise ValueError(f"No such SSM parameter {param_name}")
+        return f"{value}events"
```
+1 thanks this is much safer!

chuckwondo reviewed (approved):
Great! I like the adjustment you made to `build_app`.
